### PR TITLE
Modify bytecode representation of closures to please OCamls GC (fix #12636).

### LINKED ISSUE #12636: Potential memory corruption in VM with OCaml 4.12
Given the change of representation of closures coming in OCaml 4.12 (e.g., ocaml/ocaml#9681), all the uses of `Closure_tag` in `kernel/byterun/coq_interp.c` should be reviewed.

### PR DESCRIPTION
**Pull request by silene**
The second field of a closure can no longer be the value of the first free variable (or another closure of a mutually recursive block) but must be an offset to the first free variable.

This commit makes the bytecode compiler and interpreter agnostic to the actual representation of closures. To do so, the execution environment (variable `coq_env`) no longer points to the currently executed closure but to the last one. This has the following consequences:

- `OFFSETCLOSURE(n)` now always loads the n-th closure of a recursive block (counted from last to first);
- `ENVACC(n)` now always loads the value of the n-th free variable.

These two changes make the bytecode compiler simpler, since it no longer has to track the relative position of closures and free variables. The last change makes the interpreter a bit slower, since it has to adjust `coq_env` when executing `GRABREC`. Hopefully, cache locality will make the overhead negligible.

Fix #12636

### CODE REVIEW

**Comment by ppedrot** [comment]
@silene if you rebase this over master, we can run a bench easily.

**Comment by silene** [comment]
Once I get strong normalization to work, I will rebase. (This is driving me crazy.)

**Comment by silene** [comment]
As it happens, an unwritten invariant of the code is that closures (in particular, fixpoints) and accumulators have the same memory layout. That is, their execution environment starts at the same offset of the block. By using the new representation of closures for representing bytecode closures, this invariant broke. Arguments of partly-applied functions now start one slot further than arguments of applied accumulators. As a consequence, when strongly normalizing, the first argument of an accumulator vanishes, e.g., `fun f x => f x` gets normalized to `fun f x => f`. (Either that or irreducible fixpoints gets an extraneous 0-th argument.)

I was hoping that fixing accumulators (see also #12733) and fixing closures would be two orthogonal tasks. That was naive.

**Comment by silene** [comment]
Hopefully, the code is now correct. If so, two additional things can be done before merging the pull request:

1. Now that accumulators have the same memory layout as closures, we might just as well use the same tag. That will fix #12733 on the bytecode side.
2. Then, we can clean some mess related to native arrays. Indeed, once they no longer share the same tag as accumulator, the code can be made simpler. (`ISARRAY_INT_CAML_CALL2` can finally die.)

**Comment by silene** [comment]
@coqbot: run CI now

**Comment by silene** [comment]
@Zimmi48 Am I using the bot incorrectly?

**Comment by Zimmi48** [comment]
Hi @silene, the "run CI now" feature is a work-in-progress feature that was requested in coq/bot#33. However, the fact that no GitLab jobs were triggered for this PR is probably a bug. I'm going to look into it.

**Comment by SkySkimmer** [comment]
Until the bot's push is fixed I'll remove the single quote from the PR title.

**Comment by silene** [comment]
Here are the results of the benchmark. I am not quite sure why `coq-fourcolor` got slower; I cannot find any noticeable use of the VM (two `vm_compute` without any difference, no `vm_cast_no_check`, no `<:`, no tactics such as `ring`).

```
┌─────────────────────────────┬─────────────────────────┬─────────────────────────────────────────────┬─────────────────────────────────────────────┬───────────────────────────────┬───────────────────┐
│                             │      user time [s]      │                 CPU cycles                  │              CPU instructions               │     max resident mem [KB]     │    mem faults     │
│                             │                         │                                             │                                             │                               │                   │
│                package_name │     NEW     OLD PDIFF   │               NEW               OLD PDIFF   │               NEW               OLD PDIFF   │        NEW        OLD PDIFF   │  NEW  OLD PDIFF   │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                coq-rewriter │  620.33  625.33 -0.80 % │     1722291779222     1735598319233 -0.77 % │     2645652108550     2644550264148 +0.04 % │    1038088     987848 +5.09 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│       coq-performance-tests │ 2775.58 2791.89 -0.58 % │     7737205048947     7778809639351 -0.53 % │    14738034386377    14740254335296 -0.02 % │    4316568    4316360 +0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│ coq-rewriter-perf-SuperFast │ 1043.92 1049.91 -0.57 % │     2904462586567     2920247598250 -0.54 % │     4624953188013     4610833778931 +0.31 % │    1070688    1070768 -0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                   coq-verdi │  121.75  122.25 -0.41 % │      336820009453      337437798664 -0.18 % │      418743842826      418741992543 +0.00 % │     572732     572712 +0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│             coq-fiat-crypto │ 5086.53 5104.34 -0.35 % │    14131477968598    14179772128666 -0.34 % │    24603231643291    24391658917452 +0.87 % │    2428824    2436396 -0.31 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                 coq-unimath │ 4067.44 4081.34 -0.34 % │    11314228429016    11352692905211 -0.34 % │    21554520085190    21553005853706 +0.01 % │    1111396    1111316 +0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                  coq-geocoq │ 1492.59 1497.24 -0.31 % │     4163150844984     4170999838209 -0.19 % │     6311687074329     6310999414898 +0.01 % │    1138904    1149036 -0.88 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                   coq-flocq │  167.31  167.82 -0.30 % │      465058158500      465746885037 -0.15 % │      583907171660      583942269309 -0.01 % │     823036     822524 +0.06 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│      coq-mathcomp-odd-order │ 1080.59 1083.73 -0.29 % │     3010194396850     3019881421868 -0.32 % │     4794880685332     4794996871805 -0.00 % │     989840     989712 +0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                    coq-hott │  331.14  331.75 -0.18 % │      899725344243      903469956653 -0.41 % │     1466294843256     1466775232180 -0.03 % │     678244     678244 +0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│              coq-verdi-raft │ 1277.73 1279.69 -0.15 % │     3554534708209     3563817854740 -0.26 % │     4955387835866     4955651051275 -0.01 % │     930336     930420 -0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                coq-bedrock2 │  224.88  225.21 -0.15 % │      623385991801      623973996577 -0.09 % │     1031104077920     1028666284286 +0.24 % │    1087340    1087216 +0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│            coq-engine-bench │ 1259.50 1261.12 -0.13 % │     3527724046284     3533559074130 -0.17 % │     6066786192084     6073867146477 -0.12 % │    3994772    3996156 -0.03 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│          coq-mathcomp-field │  295.38  295.74 -0.12 % │      822660950018      822736085182 -0.01 % │     1263204449585     1263323216662 -0.01 % │     824700     826572 -0.23 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│      coq-mathcomp-character │  163.38  163.56 -0.11 % │      453978744285      455026062797 -0.23 % │      627644491267      627628660762 +0.00 % │     753112     753116 -0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                coq-compcert │  681.81  682.53 -0.11 % │     1901352076792     1903627615739 -0.12 % │     2635866414949     2633621981820 +0.09 % │    1130800    1131132 -0.03 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│             coq-lambda-rust │ 1400.91 1402.36 -0.10 % │     3901995252697     3907847032211 -0.15 % │     5547573360134     5548518935324 -0.02 % │    1119332    1120096 -0.07 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│       coq-mathcomp-solvable │  175.40  175.58 -0.10 % │      488548253901      488920978160 -0.08 % │      686951825949      687019100130 -0.01 % │     683392     683224 +0.02 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                         coq │  556.46  557.02 -0.10 % │     1557259332423     1557400273704 -0.01 % │     2131458119982     2131315035372 +0.01 % │     592936     592584 +0.06 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│      coq-mathcomp-ssreflect │   52.44   52.49 -0.10 % │      143506637165      143397999007 +0.08 % │      185060612975      185066926599 -0.00 % │     519844     519856 -0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                   coq-color │  633.80  634.10 -0.05 % │     1767974450797     1769544582250 -0.09 % │     2266583931384     2266548932525 +0.00 % │    1462524    1462420 +0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                 coq-coqutil │   60.87   60.84 +0.05 % │      168314990526      168825114800 -0.30 % │      211229198581      211211216484 +0.01 % │     533612     534408 -0.15 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│            coq-math-classes │  192.77  192.55 +0.11 % │      538808045532      539595086113 -0.15 % │      712313001038      712297294782 +0.00 % │     505284     505236 +0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│               coq-fiat-core │  108.77  108.62 +0.14 % │      309074669593      308826560475 +0.08 % │      411403686827      411386819531 +0.00 % │     485896     485936 -0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                    coq-corn │ 1718.76 1715.86 +0.17 % │     4788721538118     4778853702733 +0.21 % │     7331026080080     7330829839057 +0.00 % │     757860     758132 -0.04 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                 coq-bignums │   59.29   59.18 +0.19 % │      163663266466      164136950156 -0.29 % │      221165449685      221151090399 +0.01 % │     477156     477584 -0.09 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│        coq-mathcomp-algebra │  143.06  142.79 +0.19 % │      397846666083      397109297499 +0.19 % │      515175714296      515197945101 -0.00 % │     564076     563960 +0.02 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                coq-coqprime │   85.29   85.02 +0.32 % │      234355180220      233582529940 +0.33 % │      331198422602      330340083184 +0.26 % │     769256     769272 -0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│              coq-coquelicot │   73.34   72.99 +0.48 % │      200609920027      199929375330 +0.34 % │      259920299878      260088249541 -0.06 % │     581944     581872 +0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│            coq-fiat-parsers │  662.56  659.10 +0.52 % │     1855793260634     1845507610605 +0.56 % │     2801129806823     2779894155670 +0.76 % │    3013392    3001440 +0.40 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│       coq-mathcomp-fingroup │   49.80   49.45 +0.71 % │      137667960345      137145177502 +0.38 % │      189448008454      189451388508 -0.00 % │     485252     486140 -0.18 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│               coq-fourcolor │ 2543.26 2505.98 +1.49 % │     7090454742482     6988341710848 +1.46 % │    11677216710407    11467707991786 +1.83 % │     764260     764264 -0.00 % │    0    0  +nan % │
└─────────────────────────────┴─────────────────────────┴─────────────────────────────────────────────┴─────────────────────────────────────────────┴───────────────────────────────┴───────────────────┘
```

**Comment by proux01** [comment]
> I am not quite sure why `coq-fourcolor` got slower; I cannot find any noticeable use of the VM (two `vm_compute` without any difference, no `vm_cast_no_check`, no `<:`, no tactics such as `ring`).

I thought fourcolor was "one of the main users of vm_compute" https://github.com/coq/coq/pull/9905#issuecomment-480192565
Anyway, you were expecting a small slowdown so 1.5% sounds nice?

**Comment by silene** [comment]
> I thought fourcolor was "one of the main users of vm_compute"

My mistake. My grep for `<:` failed, while it is indeed used all over the place.

> Anyway, you were expecting a small slowdown so 1.5% sounds nice?

What I meant is that the changes to the bytecode compiler and interpreter should have a negligible overhead. (In fact, I even expect a speed up from my pull request.) But that is behind the point and I should not have mentioned it to avoid adding confusion.

The critical point is that we add one word per closure/accumulator to please the GC. In some corner cases, this might greatly increase the memory footprint, hence causing slowdown. (There is nothing specific to Coq here, as the same issue was already mentioned in the OCaml pull request.) So, 1.5% is indeed good news.

**Comment by silene** [comment]
@Zimmi48 I don't feel that confident to push it to a bugfix release. There might still be off-by-one errors in some hardly tested cases of strong normalization.

**Comment by Zimmi48** [comment]
OK thanks for the feedback. Does this mean that 8.12.1 should be marked as incompatible with OCaml 4.12?

**Comment by silene** [comment]
All versions of Coq should be marked incompatible with OCaml 4.12.

File: kernel/byterun/coq_interp.c
Comment (maximedenes) [inline kernel/byterun/coq_interp.c:962]: @bgregoir says: how about writing `index = (index == Closure_tag) ? 0 : index`

Comment (silene) [reply kernel/byterun/coq_interp.c:962]: How is that an improvement? I find it less readable.

Or perhaps you fear that the compiler cannot optimize the code otherwise? Then, fear not, as the produced assembly is the following branchless conditional on x86:

```asm
cmpq	$247, %rax
cmovne	%rax, %rcx
```

In fact, it is almost unfortunate that this conditional is optimized, because a branching conditional would be correctly predicted here. Eventually, we will have to perform a profile-guided compilation of the bytecode interpreter.

**Review by maximedenes — COMMENTED** [review]
Thanks for this nice piece of work! @bgregoir and I had a superficial look today, the principles look fine. @bgregoir wanted to have a second read this week, then we should be able to merge.

File: kernel/byterun/coq_interp.c
Comment (maximedenes) [reply kernel/byterun/coq_interp.c:962]: I think @bgregoir suggested it mainly so that it is optimized.

Comment (bgregoir) [reply kernel/byterun/coq_interp.c:962]: @silene, you are write, I want to have a CMOVXX instruction. So since the compiler use it, I'm fine with your version. 
For your remark with conditional branching and branch prediction our experiments to obtain "constant-time" code 
show also that it is generally more efficient to use CMOV instead of branching. I think it is also why the compiler use it.

Comment (silene) [reply kernel/byterun/coq_interp.c:962]: The thing is, we do not want it to be constant time, we want it to be zero time. (The fastest instruction is the one that is not executed.) As Intel's documentation states it, "Use the `SETCC` and `CMOV` instructions to eliminate unpredictable conditional branches where possible. Do not do this for predictable branches. [...] In addition, converting a conditional branch to `SETCC` or `CMOV` trades off control flow dependence for data dependence and restricts the capability of the out-of-order engine." Here, the branch is predictable when the user computes a closed term, which hopefully is the main use case for `vm_compute`.

Comment (proux01) [reply kernel/byterun/coq_interp.c:962]: Another idea would be to have a simple static analysis in the compiler detect closed terms and generate a code calling only the `NOCHECK` operations, statically saving all those tests.

Comment (bgregoir) [reply kernel/byterun/coq_interp.c:962]: I have never say that we want constant-time code ... Anyway, I am sur that using a conditional jump or a CMOV will not change the complexity of the evaluation, and since the compiler does what I want :-), I am happy with keeping this line as you want. For proux01, we have frequently discuss this solution with Maxime, we did not implement it first because you have to duplicate the code, but also because a function can be a closed term and to normalise it you have to normalise its body which is not closed anymore. I agree that this last argument is not very good, because in practice the result is a closed term in a simple datatype (without any function inside).

File: kernel/byterun/coq_interp.c
Comment (bgregoir) [inline kernel/byterun/coq_interp.c:850]: Same remark as ENVACC, for OFFSETCLOSURE.
At leat *pc++ * 3 can be done at compile time.
It is true also for  "* sizeof(value)" but I think it is less important (due to x86 and ARM addressing mode).
As before, this is a very small optimisation.

File: kernel/vmemitcodes.ml
Comment (bgregoir) [inline kernel/vmemitcodes.ml:274]: View my remark in coq_interp.c

**Review by bgregoir — COMMENTED** [review]
Thank a lot for this huge change.

File: kernel/byterun/coq_interp.c
Comment (bgregoir) [inline kernel/byterun/coq_interp.c:717]: Warning this is dangerous, and need to be carefully checked.
The risk is that block can contains a value that will be modified by the GC.
So we need to be sure that block is declare as a root when potential call to the GC (Alloc_small)

Comment (silene) [reply kernel/byterun/coq_interp.c:717]: No need, this idiom is fine. Notice that `block` is always the target of `Alloc_small` (either that or it is no longer used). So, it does not need to be marked as a root.

File: kernel/byterun/coq_interp.c
Comment (silene) [reply kernel/byterun/coq_interp.c:850]: I agree it could be done at compile time, and I hesitated to do so. But I eventually decided against it. I think it is much more important that the compilation process is agnostic to the size of closures. It is better to leave this implementation detail to the interpreter. (Same thing for `ENVACC`, but there `+ 2` comes for free.)

File: kernel/byterun/coq_interp.c
Comment (silene) [reply kernel/byterun/coq_interp.c:828]: Just a question of style. No idea why I changed it here. I suppose that, at some time, I had completely removed the line, and when I added it back later with `j+2`, I used my own style. I can change it to the style used elsewhere in the file. (That said, both styles are actually used in the file. But this might already have been my fault from an older commit.)

File: kernel/byterun/coq_interp.c
Comment (bgregoir) [reply kernel/byterun/coq_interp.c:850]: I think the most important is efficiency for the VM. Furthermore, the change is very small, since it is a trivial patch in
emitcode. I think I agree with you for ENVACC, the + 2 will come for free.

File: kernel/byterun/coq_interp.c
Comment (bgregoir) [reply kernel/byterun/coq_interp.c:828]: No need to change it.

File: kernel/byterun/coq_interp.c
Comment (maximedenes) [reply kernel/byterun/coq_interp.c:850]: > I think it is much more important that the compilation process is agnostic to the size of closures

Is there a way to have the interpreter expose this information in an explicit way? It would make it less risky (if the size of closures change in the interpreter, the corresponding constant should be updated, but the compiler does not need to be touched).

File: kernel/byterun/coq_interp.c
Comment (maximedenes) [reply kernel/byterun/coq_interp.c:828]: No need to change the patch or the code as it was before? ;)

File: kernel/byterun/coq_interp.c
Comment (bgregoir) [reply kernel/byterun/coq_interp.c:850]: Sur, the value can be exported, like for "accumulate"

File: kernel/byterun/coq_interp.c
Comment (bgregoir) [reply kernel/byterun/coq_interp.c:828]: Sorry, I mean that I do not care, so be lazy.

File: kernel/byterun/coq_interp.c
Comment (silene) [reply kernel/byterun/coq_interp.c:850]: Let us be clear. This opcode is only called during a recursive call of a mutually recursive function with three siblings or more and this recursive call is not to one of the first two siblings. So, whichever optimization we perform there, I am not even sure we can actually notice an improvement for any of the existing Coq libraries.

As for the optimization itself, it removes the following assembly instruction:
```asm
leaq    (%rax,%rax,2), %rax
```

So, even if the patch to the compiler is only a few line long, I am still not convinced it is worth skipping one micro-operation (one fourth of a cycle) in a part of the code that is almost never executed.

Comment (silene) [reply kernel/byterun/coq_interp.c:850]: That said, looking at the assembly code, I have noticed that, since `*pc` is of type `int32_t`, there are a bunch of places where the generated assembly is not as efficient as possible on a 64-bit architecture (use of `cltq` and other sign-extension mechanisms). Indeed, the compiler is trying to preserve some kind of wraparound behavior there. But I will fix it in #12959 rather than here.

Comment (bgregoir) [reply kernel/byterun/coq_interp.c:850]: Thank a lot, this code was developed before the generalisation of the 64-bit architecture.

**Comment by bgregoir** [comment]
Hi Guillaume,
   After some reflection, I think you have done two different patches in one.
   The first is changing the representation of closure, to respect the new requirement of the GC,
    this one is clearly essential.
   The second is the change in the  coq_env. 
    While it is clear that the first is needed, I am not agree with the motivation for the second.
    If I have well understood, the only motivation is to simplify a peace of code in the compiler
    but the counter part should be a small cost at execution time. 
    For me this is a bad argument: efficiency  is better that simplicity. 
    But I'm agree that simplicity without any cost is better, in particular we have to think to the next 
    developpers. Since your benches seem to show that the cost is very small, I think that maybe I'm wrong and 
    that your patch can be accepted. But I would like to do more bench before this.
    Maybe I have missed some other arguments, so do not hesitated to correct me.

**Comment by silene** [comment]
You are right, there are two patches in one. But it would have been quite painful to separate them. Indeed, it means to do the same work twice. First, when making closures compatible with OCaml 4.11, I first have to fix all the offsets to access free variables. Second, when changing `coq_env`, I again have to fix all the offsets to access free variables. Also, the most painful part was not in the compiler, it was in the decompiler used by strong normalization. That is where I spent the most time tracking off-by-one offsets.

But regarding performances, you seem to have missed the elephant in the room. Remember that free variables are accessed using the `ENVACC` opcode (and also `PUSHENVACC`, which is just `PUSH` followed by `ENVACC`). There are `ENVACC1` to `ENVACC4`, and then `ENVACC(n)`. The first ones directly encode into the opcode the offset to access, while the latter one covers the remaining cases `n > 4`. So, `ENVACC(n)` takes one more word in memory than `ENVACC1234`; and its execution requires an additional load and a bit more complicated addressing. So, whenever `ENVACC1234` can be used, there is a strict improvement to performances.

The old semantics for `ENVACC1` was to access word 1 of the closure, as given by `coq_env`. But the change to closure representation means that free variables are pushed away, so word 1 is now garbage. This means that the ratio of `ENVACC(n)` wrt `ENVACC1234` got poorer, and so did performances. But there is worse. In the case of a mutually-recursive function, not only is word 1 garbage, but so are word 2 (infix tag), word 3 (code pointer), and word 4 (GC garbage, like word 1). This is the important part, so let me stress it again. In a mutually-recursive function, only `ENVACC(n)` could be used to access free variables, because `ENVACC1234` did not point to free variables!

So, by changing the semantics of `coq_env`, I have made mutually-recursive functions as fast as plain functions. Thus, the change did not just simplify the compiler and the strong normalizer; it also made the interpreter faster, because the long form `ENVACC(n)` is now encountered much less often when executing bytecode.

By the way, in case someone wonders if the change to `coq_env` could have been performed independently from the change of closure representation, the answer is no. This was made possible only because the GC now forces us to store the offset to the free variables in word 1 of a closure. So, we might as well use this information to make the interpreter faster.

As for `OFFSETCLOSURE`'s potential inefficiency, remember that I said that this opcode can only be executed if you have a mutually-recursive function with at least three siblings (and the recursive call needs to be to the third one or later). My intuition is that there exists no such function in the wild. For instance, there is absolutely none in Coq's standard library.

**Comment by silene** [comment]
Here are some more concrete numbers that I obtained by instrumenting the compilation of the standard library. Without the change to `coq_env`, 60% of the `(PUSH)ENVACC` opcodes use the fast form. With the change to `coq_env`, 79% of the `(PUSH)ENVACC` opcodes use the fast form. Said otherwise, the number of slow opcodes is halved, thanks to the change to `coq_env`. I did not expect such a large improvement. (There are 166k `(PUSH)ENVACC` opcodes in total, so hopefully this is statistically representative.)

**Comment by bgregoir** [comment]
After talking with Guillaume, it seems not easy to have improvement to this pool request which will have a real impact on different architecture. It is frustrating to observe a slow down, but it is necessary to be compatible with Ocaml 4.11.  
So, I approve this pool request.

**Comment by silene** [comment]
Here are some points we discussed with @bgregoir that are worth keeping a record of:

1. What is the impact of the following line added to `GRABREC`? Should we have a specialized opcode to handle the case where this line does nothing interesting? (Indeed, `Int_val(Field(coq_env, 1))` is 2 in general.)
   ```c
   coq_env = coq_env + (Int_val(Field(coq_env, 1)) - 2) * sizeof(value);
   ```
   Answer: Do nothing specific. Indeed, Benjamin had a noticeable speedup by removing this line, while this was not the case for me. So, the situation seems highly dependent on the compiler (GCC vs Clang) and processor.

2. What is the impact of this pull request on the `coqprime` library? (The benchmark includes it, but it does not run its speed tests.)

   Answer: The slowdown of `vm_compute` is between 2% and 3%.

3. Is there some other way to support OCaml 4.11?

   Answer: none that we can think of. We just have to bite the bullet.

4. Is it possible to improve the compilation of `match with`? In particular, is it possible to avoid using `PUSH_RETADDR` (for non-terminal constructs) and `PUSHFIELDS`?

   Answer: This intricate compilation scheme is due to the need to strongly normalize branches, thus branches are compiled so that they be called as if they were functions. A possibility would be to compile two versions of `match with`: an optimized one that directly accesses the needed fields, and the current one for strong normalization purpose. `MAKESWITCHBLOCK` would then create an accumulator that points to the strong-normalization version rather than the optimized one that invoked it. Note that this does not require modifying the interpreter or the normalizer, only the compiler.

**Comment by Zimmi48** [comment]
Am I understanding correctly that this PR is about to be merged? If yes, I will make #12919 v8.12-only.

**Comment by silene** [comment]
Yes, it has been rebased and the bench passed.

What kind of changelog entry do you have in mind? There are no user-visible changes. And it is way too soon to say that Coq is compatible with OCaml 4.11. (Things might still change before it is released.)

**Comment by Zimmi48** [comment]
You mean OCaml 4.12. And OK, let's postpone the changelog entry claiming compatibility with OCaml 4.12 then.

**Comment by ejgallego** [comment]
We could re-enable the 4.12 CI testing if desired.

**Review by maximedenes — APPROVED** [review]
LGTM, and @bgregoir approved after discussing with @silene.

**Comment by maximedenes** [comment]
@silene Thanks a lot for this nice piece of work. I guess I should schedule some work to do the same change of representation for `native_compute`, right?

**Comment by maximedenes** [comment]
@coqbot merge now

**Comment by silene** [comment]
> I guess I should schedule some work to do the same change of representation for `native_compute`, right?

Not sure. The representation of closures also changed for native code, but contrarily to bytecode, the position of free variables was not modified. So, it depends whether `native_compute` accesses word 2 of a closure (its arity). If not, this is less pressing.

**Comment by maximedenes** [comment]
The discussion at https://github.com/coq/coq/issues/10602 seemed to suggest we had to stop using 0-tagged closures in native too. Did I misunderstand?

**Comment by silene** [comment]
That is true. But this is unrelated to the change of closure representation that comes with OCaml 4.12. This is needed for the no-naked-pointer mode of OCaml. So, that issue still needs to be fixed. But unless Ocaml's developers suddenly decide to make it the default mode for OCaml 4.12, it is less pressing.

**Comment by maximedenes** [comment]
Ok, thanks for the clarification. You're right that it is less pressing, upstream mentions they are going to stop supporting naked pointers for OCaml 5.00 (https://caml.inria.fr/pub/docs/manual-ocaml/intfc.html#ss:c-outside-head).

I also have a long-standing project of unify the first phases of native and bytecode compilation, as well as the readback (https://github.com/coq/coq/projects/17). So it would be nice to propagate the same representation changes introduced by this PR to native anyway.

**Pull request merged**
